pycaret/utils/_dependencies.py

```diff
@@ -1,15 +1,87 @@
 # Adapted from
 # https://github.com/alan-turing-institute/sktime/blob/v0.11.0/sktime/utils/validation/_dependencies.py
 
+import collections
 import sys
+from distutils.version import LooseVersion
 from importlib import import_module
-from typing import Optional
+from typing import Dict, Optional, Union
 
-from pycaret.internal.logging import get_logger
-from pycaret.utils._show_versions import _get_module_version
+from importlib_metadata import distributions
+
+from pycaret.internal.logging import get_logger, redirect_output
 
 logger = get_logger()
 
+INSTALLED_PACKAGES = None
+
+
+def _try_import_and_get_module_version(
+    modname: str,
+) -> Optional[Union[LooseVersion, bool]]:
+    """Returns False if module is not installed, None if version is not available"""
+    try:
+        if modname in sys.modules:
+            mod = sys.modules[modname]
+        else:
+            if logger:
+                with redirect_output(logger):
+                    mod = import_module(modname)
+            else:
+                mod = import_module(modname)
+        try:
+            ver = mod.__version__
+        except AttributeError:
+            # Version could not be obtained
+            ver = None
+    except ImportError:
+        ver = False
+    if ver:
+        ver = LooseVersion(ver)
+    return ver
+
+
+def get_installed_packages() -> Dict[str, Optional[LooseVersion]]:
+    global INSTALLED_PACKAGES
+    if not INSTALLED_PACKAGES:
+        # Get all installed modules and their versions without
+        # needing to import them.
+        pkg_to_dist = collections.defaultdict(list)
+        for dist in distributions():
+            for pkg in (dist.read_text("top_level.txt") or "").split():
```

Comment:
Not sure I understand what's happening here. What is "top_level.txt"?

Reply:
Added clarification, PTAL